freedomofpress/securedrop-workstation
# Add sd-export VMs and basic export flow

### Description
Towards #84 :
This adds:
* `sd-export-template`, a TemplateVM to manage every package related to (all?) exports
* sd-export-usb-dvm DispVM template: a Disposable VM
* sd-export-usb named disposable VM (that will be used once per session). 

USB devices connected to a given USB port will automatically be attached to `sd-export-usb`, and `sd-export-usb` has a mime handler that will automatically copy .gpg files to the attached USB storage.

There are a few limitations in this current approach:

1. Dedicated USB port for exporting. It seems like that would less error prone than using the Qubes menus, if there's some sort of labeling on the workstation itself.

2. Transfer device provisioning is not automatic, it may be difficult to do without risking the data.

3. Export staging area (Nautilus window in `sd-export-usb`) would provide more user flexibility, but also more complexity.

4. Error handling/feedback may be challenging with the `qvm-open-in-vm` approach

### Test Plan
- Plug in the usb stick and update `config.json` per the docs
- Create an encrypted transfer device (ext4-luks encrypted volume)
- [ ] `make all` and `make test` or `make sd-export` and `make test` complete successfully and all tests pass
- Create a tarball (important, extension must be `.sd-export`) in `sd-svs` containing the file structure described in the docs.
- [ ] In sd-svs, run `qvm-open-in-vm sd-export-usb <file>.sdexport` completes without error
- [ ] The file was successfully copied to the disk
- [ ] Ensure `sd-export-usb`  is indeed disposable (for example, home directory contents don't persist reboots)


Note that the first iteration (https://github.com/freedomofpress/securedrop-workstation/pull/259/commits/85f3d67d3e3a8651eafa75183a7b1ed360778eb6) was using non-encrypted devices, with symmetrically-encrypted files. Due to the default behavior or most tools extracting and storing the plaintext version next to the encrypted version on the filesystem, using a disk that was not full disk encrypted is likely not a very good idea, should the USB be misplaced.

### Demo

![sd-export-small](https://user-images.githubusercontent.com/15223328/58722135-ad7c7200-83a4-11e9-9efa-ac92a27ce1f6.gif)
